mantidproject/mantid
# Remove GSL from BSpline

**Description of work.**

This PR removes the use of GSL from the `BSpline` function and replaces it with Eigen.
It also introduces Function Attribute Validators to `BSpline`.

The ability to add 'error' to bounded validators to help deal with floating point arithmetic errors has been provided.

A related bug has been fixed in `FunctionTreeView.cpp` that caused vector attributes to be set (triggering the validator) before they had been populated.

Also, more minor change to remove `GSL` from `TrustRegion`

1. Observe that the `LevenbergMarquardtMDTest`, and `BSplineTest` unit tests pass.
2. Observe all other fitting related unit tests pass
3. Test fitting a `BSpline` to a spectra:
      - Load the `MAR11060` dataset into Mantid (from training data)
      - Plot spectrum `2`
      - Open the `FitPropertyBrowser` by clicking `fit` on the plot window
      - Add a function by right clicking on functions header
      - Select `background` -> `BSpline` -> `Ok`
      - Expand the newly added function, play around with the function attributes to test validation:
            - `StartX` cannot be greater than `EndX`, and vice versa
            - Breakpoints must be between `StartX` and `EndX`, inclusive
            - `Order` must be greater than 0.
            - `NBreak` must be greater than 2.
      - Increase `NBreak` to 14
      - Increase `Order` to 4.
      - Click `fit`, observe `Changes in function value are too small`
      - Uncheck `uniform`
      - Change the first 9 Breakpoints to: `105.182734`, `1000`, `1500`, `2000`, `2500`, `3000`, `4000`, `5000`, `6000`
      - Click `fit`, observe `success`

Towards resolving #34278.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
